### PR TITLE
Support gormio

### LINKED ISSUE #810: Add instrumentation for gorm v2
Gorm v2 has significant performance improvements and features.

More notably it has context support by default.

https://gorm.io/docs/v2_release_note.html#Major-Features

### PR DESCRIPTION
**Pull request by Deepak13245**
Resolves #810

### CODE REVIEW

**Comment by axw** [comment]
@Deepak13245 thanks for the PR!

Looks pretty good, but I'm wondering if the callbacks are still necessary. If context is now first class, then perhaps we can instead just rely on `apmsql` to create the spans?

Also, I'd like to call this module `apmgormv2` to be more in line with other modules.

**Comment by Deepak13245** [comment]
@axw renamed to apmgormv2
regarding the creation of spans by apmsql, gorm doesn't accept *sql.DB directly in params. There is a workaround so I raised a separate PR https://github.com/Deepak13245/apm-agent-go/pull/1. Please suggest if there is a better approach.

Thanks

**Comment by axw** [comment]
A couple of things bother me about the plugin approach:
 - It closes and reopens the connection. Not the end of the world, just a bit annoying.
 - As it is at the moment, extractDsn forces a dependency on all the drivers. We could solve this by having the "_" dialect imports register a function though.

What do you think about providing "Open" functions inside the dialect packages instead? So instead of

```go
import (
    "gorm.io/driver/sqlite"
    "gorm.io/gorm"         
)

db, err := gorm.Open(sqlite.Open("gorm.db"), &gorm.Config{})
```

you could write

```go
import (
    "go.elastic.co/apm/module/apmgormv2/dialect/apmsqlite"
    "gorm.io/gorm"
)

db, err := gorm.Open(apmsqlite.Open("gorm.db"), &gorm.Config{})
```

Internally this would use the recently introduced support for custom drivers: https://github.com/go-gorm/sqlite/pull/11

Basically just:

```go
package apmsqlite // go.elastic.co/module/apmgormv2/dialect/apmsqlite

import (
    "gorm.io/driver/sqlite"
    "gorm.io/gorm"
    "go.elastic.co/module/apm/module/apmsql"
    _ "go.elastic.co/module/apm/module/apmsql/sqlite3"
)

func Open(dsn string) gorm.Dialector {
        return &sqlite.Dialector{
                DriverName: apmsql.DriverPrefix + sqlite.DriverName,
                DSN:        dsn,
        }
}
```

**Comment by Deepak13245** [comment]
@axw creating dialect from with driver prefix makes it neat, updated the pr please check.

File: module/apmgormv2/driver/sqlite/init.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+	_ "go.elastic.co/apm/module/apmsql/sqlite3" // register sqlite3 with apmsql
+)
+
+// Open creates a dialect with apmsql
+func Open(dsn string) gorm.Dialector {
+	driverName := sqlite.Dialector{}.Name()
+
+	dialect := &sqlite.Dialector{
+		DSN: dsn,
+		DriverName: driverName,
```
Comment (axw) [inline module/apmgormv2/driver/sqlite/init.go:35]:
```suggestion
		DriverName: apmsql.DriverPrefix + driverName,
```

File: module/apmgormv2/driver/postgres/init.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+// Package apmpostgres imports the gorm postgres dialect package,
+// and also registers the lib/pq driver with apmsql.
+package apmgpostgres
+
+import (
+	"go.elastic.co/apm/module/apmsql"
+	"gorm.io/driver/postgres"
+	"gorm.io/gorm"
+
+	_ "go.elastic.co/apm/module/apmsql/pq" // register lib/pq with apmsql
```
Comment (axw) [inline module/apmgormv2/driver/postgres/init.go:27]: Apparently gorm v2 uses pgx by default: https://github.com/go-gorm/postgres/blob/master/postgres.go

Maybe don't add the postgres dialect for now? WDYT?

**Review by axw — COMMENTED** [review]
Thanks! I agree this looks much neater.

**Comment by axw** [comment]
jenkins run the tests please

File: module/apmgormv2/driver/postgres/init.go (hunk 2 above)
Comment (Deepak13245) [reply module/apmgormv2/driver/postgres/init.go]: Yes, Removing Postgres dialect for now

**Review by axw — APPROVED** [review]
Perfect, thanks!

**Comment by axw** [comment]
jenkins run the tests please

**Comment by axw** [comment]
> [2020-10-07T02:40:59.696Z] ../../gorm.io/gorm/utils/utils.go:46:30: reflect.ValueOf(val).IsZero undefined (type reflect.Value has no field or method IsZero)

It looks like gorm v2 requires Go 1.13+, so we'll need to update the build constraint.

**Comment by axw** [comment]
Looks good, I think the remaining test failures are fixed on master. Can you please merge master so we can make sure?

File: module/apmgormv2/doc.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// +build go1.14
+
```
Comment (axw) [inline module/apmgormv2/doc.go:19]:
```suggestion
```

One last thing I hope. This is a bit of a hack to get `check-vet` to stop failing for older versions of Go.

**Comment by axw** [comment]
CI looks happy. Thanks for another nice PR, and for bearing with my requests :)
